File: include/GUIObj.r
#ifndef _GUIOBJ_R_
#define _GUIOBJ_R_

#include "Class.r"

#define CONT_GUIOBJ \
	CONT_CLASS \
	int (*update) ( void *obj ); \
	void (*draw) ( void *obj );

struct GUIObj {
	CONT_GUIOBJ
};

#endif

File: src/GUIObj.c
#include "GUIObj.r"
#include "GUIObj.h"

#include <assert.h>
#include "debug.h"

int update ( void *_self ) {
	const struct GUIObj **self = _self;
	int ret = 0;

	assert ( *self );
	if ( (*self)->update ) {
		ret = (*self)->update ( _self );
	}

	return ret;
}

void draw ( void *_self ) {
	const struct GUIObj **self = _self;

	assert ( *self );
	if ( (*self)->draw ) {
		(*self)->draw ( _self );
	}
}
